pkg/virt-handler/device-manager/usb_device.go

```diff
@@ -0,0 +1,547 @@
+			continue
+		}
+		switch values[0] {
+		case "BUSNUM":
+			val, err := strconv.ParseInt(values[1], 10, 32)
+			if err != nil {
+				return nil
+			}
+			u.Bus = int(val)
+		case "DEVNUM":
+			val, err := strconv.ParseInt(values[1], 10, 32)
+			if err != nil {
+				return nil
+			}
+			u.DeviceNumber = int(val)
+		case "PRODUCT":
+			products := strings.Split(values[1], "/")
+			if len(products) != 3 {
+				return nil
+			}
+
+			val, err := strconv.ParseInt(products[0], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.Vendor = int(val)
+
+			val, err = strconv.ParseInt(products[1], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.Product = int(val)
+
+			val, err = strconv.ParseInt(products[2], 16, 32)
+			if err != nil {
+				return nil
+			}
+			u.BCD = int(val)
+		case "DEVNAME":
+			u.DevicePath = "/dev/" + values[1]
+		default:
+			log.Log.V(5).Infof("Skipping unhandled line: %s", line)
+		}
+	}
+	return &u
+}
+
+type LocalDevices struct {
+	devices []*USBDevice
+}
+
+// finds by vendor and product
+func (l *LocalDevices) find(vendor, product int) *USBDevice {
+	for _, local := range l.devices {
+		if local.Vendor == vendor && local.Product == product {
+			return local
+		}
+	}
+	return nil
+}
+
+// remove all cached elements
+func (l *LocalDevices) remove(cache map[string]*USBDevice) {
+	devices := []*USBDevice{}
+	for _, local := range l.devices {
+		if _, exists := cache[local.GetID()]; !exists {
+			devices = append(devices, local)
+		}
+	}
+	l.devices = devices
+}
+
+func discoverPluggedUSBDevices() *LocalDevices {
+	usbDevices := make([]*USBDevice, 0)
+	err := filepath.Walk(PathToUSBDevices, func(path string, info os.FileInfo, err error) error {
+		// Ignore named usb controllers
+		if strings.HasPrefix(info.Name(), "usb") {
+			return nil
+		}
+		// We are interested in actual USB devices information that
+		// contains idVendor and idProduct. We can skip all others.
+		if _, err := os.Stat(filepath.Join(path, "idVendor")); err != nil {
+			return nil
+		}
+
+		// Get device information
+		if device := parseSysUeventFile(path); device != nil {
+			usbDevices = append(usbDevices, device)
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.Log.Reason(err).Error("Failed when walking usb devices tree")
+	}
+	return &LocalDevices{devices: usbDevices}
+}
+
+func parseSelector(s *v1.USBSelector) (int, int, error) {
+	val, err := strconv.ParseInt(s.Vendor, 16, 32)
+	if err != nil {
+		return -1, -1, err
+	}
+	vendor := int(val)
+
+	val, err = strconv.ParseInt(s.Product, 16, 32)
+	if err != nil {
+		return -1, -1, err
+	}
+	product := int(val)
+
+	return vendor, product, nil
+}
+
+func discoverAllowedUSBDevices(usbs []v1.USBHostDevice) map[string][]*PluginDevices {
+	// The return value: USB Device Plugins found and permitted to be exposed
+	plugins := make(map[string][]*PluginDevices)
+	// All USB devices found plugged in the Node
+	localDevices := discoverPluggedUSBDevices()
+	for _, usbConfig := range usbs {
```

Comment:
As already discussed offline, this is a quite complex loop. You could try to refactor and maybe simplify it using maps.